### PR TITLE
lbar-439 : Administration OPCO

### PR DESCRIPTION
**Pull request by kevbarns**
Interface d'administration pour les utilisateurs de type `OPCO` 

- Intégration complète de React-table, avec barre de recherche et export des données dans de nouveau composant (TableNew & ExportButtonNew). 
- Va remplacer ReactiveSearch ultérieurement.
- Fonction d'export non générique pour le moment.
- Mise à jour de la version de MongoDB pour accéder au fonction d'aggregation dans des requêtes find() [$expr](https://www.mongodb.com/docs/manual/reference/operator/query/expr/)

### CODE REVIEW

**Review by alanlr — APPROVED** [review]

File: server/src/jobs/user/misc/addOpcoInformationToEntreprise.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+runScript(async () => {
+  const entreprises = await User.find({ type: "ENTREPRISE", opco: { $exists: false } }).lean();
+
+  logger.info(`${entreprises.length} etp à mettre à jour...`);
+
+  await asyncForEach(entreprises, async (etp) => {
+    const form = await Formulaire.findOne({ siret: etp.siret });
+
+    if (form?.opco) {
+      // logger.info(`updating ${etp.siret} : opco: ${form.opco} — idcc: ${form.idcc} `);
```
Comment (rimiti) [inline server/src/jobs/user/misc/addOpcoInformationToEntreprise.js:15]: Are you sure to want to comment this log?

File: ui/src/components/ConfirmationDesactivationUtilisateur.jsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  useToast,
+} from '@chakra-ui/react'
+import { useState } from 'react'
+import { useQueryClient } from 'react-query'
+import { updateUserValidationHistory } from '../api'
+import useAuth from '../common/hooks/useAuth'
+import { Close } from '../theme/components/icons'
+
+export default (props) => {
+  let { isOpen, onClose, raison_sociale, _id } = props
```
Comment (rimiti) [inline ui/src/components/ConfirmationDesactivationUtilisateur.jsx:23]:
```suggestion
  const { isOpen, onClose, raison_sociale, _id } = props
```

File: ui/src/components/ConfirmationDesactivationUtilisateur.jsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  const toast = useToast()
+  const client = useQueryClient()
+  const [motif, setMotif] = useState()
+  const [auth] = useAuth()
+
+  const disableUser = () => {
```
Comment (rimiti) [inline ui/src/components/ConfirmationDesactivationUtilisateur.jsx:29]: jsdoc is missing on all of your functions.

File: ui/src/components/ExportButton/ExportButtonNew.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import { Button, Link } from '@chakra-ui/react'
+import dayjs from 'dayjs'
```
Comment (rimiti) [inline ui/src/components/ExportButton/ExportButtonNew.js:2]: 🎖️

File: ui/src/components/ExportButton/ExportButtonNew.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import _ from 'lodash'
```
Comment (rimiti) [inline ui/src/components/ExportButton/ExportButtonNew.js:3]: Directly importing the right function is more lightweight :)

File: ui/src/components/ExportButton/ExportButtonNew.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  const format = data
+    .flat()
+    .map((x) => _.omit(x, ['_id', 'etat_utilisateur', '__v', 'isAdmin', 'type', 'scope', 'email_valide', 'id_form']))
+    .map((x) => {
+      return {
+        ...x,
+        createdAt: dayjs(x.createdAt).format('YYYY-MM-DD'),
+        updatedAt: dayjs(x.updatedAt).format('YYYY-MM-DD'),
+      }
+    })
```
Comment (rimiti) [inline ui/src/components/ExportButton/ExportButtonNew.js:17]:
```suggestion
    .map((x) => ({
        ...x,
        createdAt: dayjs(x.createdAt).format('YYYY-MM-DD'),
        updatedAt: dayjs(x.updatedAt).format('YYYY-MM-DD'),
      })
```

File: ui/src/components/ExportButton/ExportButtonNew.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    .map((x) => {
+      return { ...x, siret: `"${x.siret}"`, telephone: `"${x.telephone}"` }
+    })
```
Comment (rimiti) [inline ui/src/components/ExportButton/ExportButtonNew.js:20]:
```suggestion
    .map((x) => ({ ...x, siret: `"${x.siret}"`, telephone: `"${x.telephone}"` }))
```

File: ui/src/components/ExportButton/ExportButtonNew.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  let fileName = `${datasetName}_${new Date().toJSON()}.csv`
```
Comment (rimiti) [inline ui/src/components/ExportButton/ExportButtonNew.js:21]:
```suggestion
  const fileName = `${datasetName}_${new Date().toJSON()}.csv`
```

File: ui/src/components/InformationLegaleEntreprise.jsx [hunk 9]
```diff
@@ -7,20 +7,19 @@ import InfoTooltip from './InfoToolTip'
...
-    libelle_naf,
     tranche_effectif,
     date_creation_etablissement,
     type,
     adresse,
     qualiopi,
   } = props
-  const RAISON_SOCIALE = raison_sociale.length > 30 ? raison_sociale.substring(0, 30) + '...' : raison_sociale
+
+  const RAISON_SOCIALE = raison_sociale.length > 30 ? raison_sociale.substring(0, 30) + '...' : raison_sociale ?? ''
```
Comment (rimiti) [inline ui/src/components/InformationLegaleEntreprise.jsx:22]: This one liner is a bit complex to understand.

File: ui/src/components/Layout/NavigationMenu.jsx [hunk 10]
```diff
@@ -64,23 +64,11 @@ const NavLinks = ({ isOpen }) => {
...
         textStyle='sm'
       >
-        {auth.sub === 'anonymous' && (
-          <>
-            <NavItem to='/'>Accueil</NavItem>
-          </>
-        )}
-
+        <NavItem to='/'>Accueil</NavItem>
+        {auth.type === 'OPCO' && <NavItem to={`/administration/opco`}>Entreprises</NavItem>}
```
Comment (rimiti) [inline ui/src/components/Layout/NavigationMenu.jsx:68]: Could be nice to have an object like

```js
const authType = {
OPCO: 'OPCO',
CFA: 'CFA',
'ENTREPRISE': 'ENTREPRISE'
}
```

Then, you'll can do authType.OPCO. :)

Comment (rimiti) [reply ui/src/components/Layout/NavigationMenu.jsx:68]: This will avoid any typo

File: ui/src/components/PaginationReactQuery.jsx [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+      leftIcon={<ChevronLeft />}
+      as={Link}
+      variant='link'
+      isDisabled={!canPreviousPage}
+      onClick={() => previousPage()}
+    >
+      Précédent
+    </Button>
+  )
+}
```
Comment (rimiti) [inline ui/src/components/PaginationReactQuery.jsx:25]:
```suggestion
const PreviousLink = ({ previousPage, canPreviousPage }) => 
    <Button
      leftIcon={<ChevronLeft />}
      as={Link}
      variant='link'
      isDisabled={!canPreviousPage}
      onClick={() => previousPage()}
    >
      Précédent
    </Button>
```

File: ui/src/components/PaginationReactQuery.jsx [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    <Button rightIcon={<ChevronRight />} as={Link} variant='link' isDisabled={!canNextPage} onClick={() => nextPage()}>
+      Suivant
+    </Button>
+  )
+}
+
+export default ({ page, canPreviousPage, canNextPage, pageCount, nextPage, previousPage, gotoPage, currentPage }) => {
+  if (Number.isNaN(pageCount) || pageCount <= 1) {
+    return <></>
+  }
```
Comment (rimiti) [inline ui/src/components/PaginationReactQuery.jsx:38]: Isn't it better to return null instead a fragment?

File: ui/src/components/PaginationReactQuery.jsx [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+  if (pageCount <= 5) {
+    return (
+      <Box className='search-pagination' textAlign='center' my={3} mx={1}>
+        <PreviousLink previousPage={previousPage} canPreviousPage={canPreviousPage} />
+        {Array(pageCount)
+          .fill(page)
+          .map((_unusedValue, index) => {
+            return <PageLink />
+          })}
```
Comment (rimiti) [inline ui/src/components/PaginationReactQuery.jsx:48]:
```suggestion
          .map((_unusedValue, index) => <PageLink />)
```

File: ui/src/components/PaginationReactQuery.jsx [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+        <PageLink pageNumber={currentPage + 4} gotoPage={gotoPage} />
+      )}
+
+      {currentPage + 2 < pageCount - 1 && <span>...</span>}
+
+      {currentPage < pageCount - 1 && <PageLink pageNumber={pageCount} gotoPage={gotoPage} />}
+
+      <NextLink nextPage={nextPage} canNextPage={canNextPage} />
+    </Box>
+  )
```
Comment (rimiti) [inline ui/src/components/PaginationReactQuery.jsx:86]: Piouf, seems black magic :)

File: ui/src/components/TableNew.jsx [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+        border='none'
+        pointerEvents='none'
+        children={<SearchLine color='white' />}
+      />
+    </InputGroup>
+  )
+}
+
+function fuzzyTextFilterFn(rows, id, filterValue) {
+  console.log('fuzzy', rows)
```
Comment (rimiti) [inline ui/src/components/TableNew.jsx:47]: This logs seems to be removable

File: ui/src/components/TableNew.jsx [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+  return matchSorter(rows, filterValue, { keys: [(row) => row.values[id]], threshold: matchSorter.rankings.CONTAINS })
+}
+
+// Let the table remove the filter if the string is empty
+fuzzyTextFilterFn.autoRemove = (val) => !val
+
+export default ({ data, columns, description }) => {
```
Comment (rimiti) [inline ui/src/components/TableNew.jsx:54]: adding jsdoc to explain a bit this function could ne nice.

File: ui/src/pages/Administration/Account.jsx [hunk 17]
```diff
@@ -13,20 +13,53 @@ import {
...
 export default () => {
   const toast = useToast()
   const [auth] = useAuth()
   const navigate = useNavigate()
   const client = useQueryClient()
+  const params = useParams()
+
+  const getUserDataFetchingContext = () => {
+    switch (auth.type) {
+      case 'OPCO':
```
Comment (rimiti) [inline ui/src/pages/Administration/Account.jsx:38]: You can reuse your authType.OPCO here

**Review by rimiti — APPROVED** [review]
PR a bit complicated to review but well done.

LGTM

File: server/src/jobs/user/misc/addOpcoInformationToEntreprise.js (hunk 1 above)
Comment (kevbarns) [reply server/src/jobs/user/misc/addOpcoInformationToEntreprise.js:15]: yes it's no useful. I could even remove it actually

**Pull request merged**